src/coreclr/src/inc/corprof.idl

```diff
@@ -4019,6 +4066,39 @@ interface ICorProfilerInfo11 : ICorProfilerInfo10
                 [in, string] const WCHAR *szValue);
 }
 
+[
+    object,
+    uuid(27b24ccd-1cb1-47c5-96ee-98190dc30959),
+    pointer_default(unique),
+    local
+]
+interface ICorProfilerInfo12 : ICorProfilerInfo11
+{
+    HRESULT EventPipeCreateProvider(
+                [in, string] const WCHAR    *szName,
+                [out] EVENTPIPE_PROVIDERID  *pProviderID);
+
+    HRESULT EventPipeDefineEvent(
+                [in] EVENTPIPE_PROVIDERID   provHandle,
+                [in, string] const WCHAR    *szName, 
+                [in] UINT32                 eventID,
+                [in] UINT64                 keywords,
+                [in] UINT32                 eventVersion,
+                [in] UINT32                 level,
+                [in] BOOL                   needStack,
+                [in] UINT32                 cParamDescs,
+                [in, size_is(cParamDescs)] 
+                     COR_PRF_EVENTPIPE_PARAM_DESC pParamDescs[],
+                [out] EVENTPIPE_EVENTID     *pEventID);
+
+    HRESULT EventPipeWriteEvent(
+                [in] EVENTPIPE_EVENTID  event,
+                [in] BYTE               *pData,
```

Comment:
What do you think about changing this to an array of arguments and making the runtime responsible for packing it? We already have the code to do that [here](https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/coreclr/src/vm/eventpipe.cpp#L525)

Reply:
I thought about that but am worried about the perf impact. The profiler would have to do an unnecessary allocation and fill out metadata for every event. Given how easy the serializtion to a byte * is, it doesn't seem worth it to me. I'm not extremely against it though, I could be talked in to the way you suggest.

Reply:
> The profiler would have to do an unnecessary allocation

I would anticipate having the runtime pack it has the same or fewer (heap) allocations. As an example assume we've got an event that takes an int and a string as parameters.

Runtime packs it:
````
CORPROF_EVENTDATA args[2];
args[0].Ptr = &intParam;
args[0].Size = 4;
args[1].Ptr = pStringParam;
args[1].Size = strlen(pStringParam);
return pProfiler->WriteEvent(event, args, 2, …)
````

Profiler packs it:
````
DWORD stringLen = strlen(pStringParam);
DWORD bufferSize = 4 + 2+ stringLen;
if(stringLen > 0xFFFF) return E_FAIL;
BYTE* pBuffer = new (nothrow) BYTE[bufferSize]; // this is an extra allocation we can't optimize away
if(pBuffer == nullptr) return E_FAIL;
memcpy(pBuffer, &intParam, 4);
*(UINT16*)(pBuffer+4) = stringLen;
memcpy(pBuffer+6, pStringParam, stringLen);
HRESULT ret = pProfiler->WriteEvent(event, pBuffer, bufferSize, ...)
delete [] pBuffer;
return ret;
````

Aside from perf, two other reasons to favor runtime packing:
a) Flexibility to change the packing format. If we ever have an updated or alternate serialization format we would need to deserialize and reserialize the payload buffer.
b) Reduce complexity/chance for mistakes in the profiler. I doubt its going to obvious how to pack a string for example. I am not sure if I what I wrote by memory is correct and if it is it also took almost twice as much code to do the same task.

Reply:
Ok, you convinced me 😄